src/Foundatio.Extensions.Hosting/Jobs/ScheduledJobRunner.cs

```diff
@@ -0,0 +1,113 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Cronos;
+using Foundatio.Caching;
+using Foundatio.Jobs;
+using Foundatio.Lock;
+using Foundatio.Utility;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Foundatio.Extensions.Hosting.Jobs;
+
+internal class ScheduledJobRunner
+{
+    private readonly Func<IServiceProvider, IJob> _jobFactory;
+    private readonly IServiceProvider _serviceProvider;
+    private CronExpression _cronSchedule;
+    private readonly ILockProvider _lockProvider;
+    private readonly ILogger _logger;
+    private readonly DateTime _baseDate = new(2010, 1, 1);
+    private string _cacheKeyPrefix;
+
+    public ScheduledJobRunner(string schedule, string jobName, Func<IServiceProvider, IJob> jobFactory, IServiceProvider serviceProvider, ICacheClient cacheClient, ILoggerFactory loggerFactory = null)
+    {
+        _jobFactory = jobFactory;
+        _serviceProvider = serviceProvider;
+        Schedule = schedule;
+        JobName = jobName;
+        _logger = loggerFactory?.CreateLogger<ScheduledJobRunner>() ?? NullLogger<ScheduledJobRunner>.Instance;
+
+        _cronSchedule = CronExpression.Parse(schedule);
+        if (_cronSchedule == null)
+            throw new ArgumentException("Could not parse schedule.", nameof(schedule));
+
+        var interval = TimeSpan.FromDays(1);
+
+        var nextOccurrence = _cronSchedule.GetNextOccurrence(SystemClock.UtcNow);
+        if (nextOccurrence.HasValue)
+        {
+            var nextNextOccurrence = _cronSchedule.GetNextOccurrence(nextOccurrence.Value);
+            if (nextNextOccurrence.HasValue)
+                interval = nextNextOccurrence.Value.Subtract(nextOccurrence.Value);
+        }
+
+        _lockProvider = new ThrottlingLockProvider(cacheClient, 1, interval.Add(interval));
+
+        NextRun = _cronSchedule.GetNextOccurrence(SystemClock.UtcNow);
+    }
+
+    private string _schedule;
+    public string Schedule
+    {
+        get { return _schedule;}
+        set
+        {
+            _cronSchedule = CronExpression.Parse(value);
+            NextRun = _cronSchedule.GetNextOccurrence(SystemClock.UtcNow);
```

Comment:
Should you be able to change the schedule of anything in flight? What if they do a cron schedule every second?

Reply:
Cron schedules can only be every minute at the most.